### ISSUE TITLE
Linter set to ES5

### ISSUE THREAD

**Issue by NovemLinguae** [wikimedia-gadgets/twinkle#1515]
I notice our linter is set to enforce ES5 and reject any JavaScript features newer than that. This is a little inconvenient to the developer, who has to avoid newer language features such as match, let, class, await, startsWith, endsWith, includes, regex s flag, etc.

Was just wondering what the idea behind keeping the linter on ES5 is. I know [Twinkle has 45,000 users](https://en.wikipedia.org/wiki/Special:GadgetUsage) or so. I guess the assumption is that a small subset of these folks are using very old browsers that don't support ES6+?

Is it worth the tradeoff? Should we plan to allow ES6 sometime in the future?

**Comment by siddharthvp** [comment]
The reason is just that MW gadgets don't yet support ES6. We need https://phabricator.wikimedia.org/T75714 to be resolved first one way or another. It is quite hard to resolve satisfactorily - as it requires introducing a ES6-capable validator. I have a [rather hacky patch](https://gerrit.wikimedia.org/r/c/mediawiki/extensions/Gadgets/+/758086) filed as an interim solution.

**Comment by NovemLinguae** [comment]
[This patch](https://gerrit.wikimedia.org/r/c/mediawiki/extensions/Gadgets/+/758086/) was +2'd today, opening the door for gadgets to start using ES6. Next Thursday when this deploys, do we want to switch Twinkle over to ES6?

If so, here is our todo list:

[ ] Make a post at the interface administrator's noticeboard requesting this change
[ ] Reconfigure our linter to allow ES6

**Comment by siddharthvp** [comment]
I am on board with this. If @MusikAnimal is also in agreement, @NovemLinguae can you please file the patch for updating the linter reconfiguration, [gadget definition file](https://github.com/wikimedia-gadgets/twinkle/blob/master/gadget.txt) and the README? On-wiki gadget definition change can be handled as part of deployment and IMO doesn't need an IANB post.

**Comment by MusikAnimal** [comment]
100% in support! We may kill IE support in doing this, but nowadays that's the intention: https://www.mediawiki.org/wiki/Compatibility/IE11

I'm happy to do the deployment when the time comes. Just ping me.

**Comment by NovemLinguae** [comment]
I see a bunch of JavaScript versions here. Are these all safe for Twinkle/gadget use now? Or do we need to look into having the linter continue enforcing a block of some of the newer ones?

<img width="435" alt="image" src="https://user-images.githubusercontent.com/79697282/197303619-108579bf-2a87-453a-9470-9efa771de715.png">

**Comment by siddharthvp** [comment]
We're still limited by what the [minifier](https://github.com/wikimedia/minify) supports, which is ES6 == ES2015 (see https://phabricator.wikimedia.org/T277675). However, we're still able to use Array.includes(), Object.values(), etc which are merely new methods rather than new syntax.

**Comment by NovemLinguae** [comment]
On Thursday, I think we want to change https://en.wikipedia.org/wiki/MediaWiki:Gadgets-definition

```
Twinkle[ResourceLoader|dependencies=ext.gadget.morebits,ext.gadget.select2,mediawiki.api,mediawiki.language|rights=autoconfirmed|type=general|peers=Twinkle-pagestyles]|Twinkle.js|Twinkle.css|twinklearv.js|twinklewarn.js|twinkleblock.js|friendlywelcome.js|friendlyshared.js|friendlytalkback.js|twinklespeedy.js|twinkleprod.js|twinklexfd.js|twinkleimage.js|twinkleprotect.js|friendlytag.js|twinklediff.js|twinkleunlink.js|twinklefluff.js|twinkledeprod.js|twinklebatchdelete.js|twinklebatchprotect.js|twinklebatchundelete.js|twinkleconfig.js
```

to

```
Twinkle[ResourceLoader|dependencies=ext.gadget.morebits,ext.gadget.select2,mediawiki.api,mediawiki.language|rights=autoconfirmed|type=general|peers=Twinkle-pagestyles**|requiresES6**]|Twinkle.js|Twinkle.css|twinklearv.js|twinklewarn.js|twinkleblock.js|friendlywelcome.js|friendlyshared.js|friendlytalkback.js|twinklespeedy.js|twinkleprod.js|twinklexfd.js|twinkleimage.js|twinkleprotect.js|friendlytag.js|twinklediff.js|twinkleunlink.js|twinklefluff.js|twinkledeprod.js|twinklebatchdelete.js|twinklebatchprotect.js|twinklebatchundelete.js|twinkleconfig.js
```

Right?

**Comment by siddharthvp** [comment]
Yes

**Comment by MusikAnimal** [comment]
[Done](https://en.wikipedia.org/w/index.php?title=MediaWiki:Gadgets-definition&diff=prev&oldid=1118523961&diffmode=source)!

**Comment by NovemLinguae** [comment]
@MusikAnimal. Thank you my friend. Can you also set |requiresES6 for Morebits.js? I think we decided to do that in the pull request.

**Comment by MusikAnimal** [comment]
> Can you also set |requiresES6 for Morebits.js? I think we decided to do that in the pull request.

Is that going to break any other scripts that use Morebits? As in, after we deploy any ES6-only code? I know this isn't MediaWiki proper, but [libraries shouldn't normally be ES6](https://www.mediawiki.org/wiki/ResourceLoader/ES6#When_it's_OK_to_use_ES6) (for now). Maybe we should a Babel build step for Morebits

**Comment by siddharthvp** [comment]
The issue would be that we would need a separate lint config for morebits, and I don't think eslint supports file-specific override for ecmaVersion (though I suppose we could put morebits in a subdir and put another .eslintrc in there). 

Morebits isn't much used outside Twinkle. I think the biggest use is DYK-helper with 340 users (which is written by myself - and I'm fine with not having IE11 support for that), the next biggest use is [arb.js](https://en.wikipedia.org/wiki/User:Bellezzasolo/Scripts/arb.js) with 70 users, which is dependent on Twinkle so won't work on IE11 anyway with Twinkle requiring ES6. So I think there's little point in retaining IE11 support in morebits and the costs are not justified.

**Comment by MusikAnimal** [comment]
There's a few more I found with [Global Search](https://global-search.toolforge.org/?q=Morebits%5C.&regex=1&namespaces=2%2C4%2C8&title=%28Gadgets-definition%7C.*%5C.%28js%7Ccss%7Cjson%29%29) (only enwiki listed below, but there are other results, assuming the ES6 migration will eventually carry over cross-wiki):
* https://en.wikipedia.org/wiki/User:CAPTAIN%20MEDUSA/Redirect%20Helper.js
* https://en.wikipedia.org/wiki/User:Abelmoschus%20Esculentus/rater.js
* https://en.wikipedia.org/wiki/User:Ethanbas/scripts/AFCRHS.js
* https://en.wikipedia.org/wiki/User:Lee%20Vilenski/FAC-helper.js
* https://en.wikipedia.org/wiki/User:Jutyaar/common.js
* https://en.wikipedia.org/wiki/User:This,%20that%20and%20the%20other/ratemath.js
* https://en.wikipedia.org/wiki/User:ST47/common.js
* …

Some of these seem to be custom extensions to Twinkle/Morebits, which may mean they were fragile or prone to breakage to begin with. I didn't thoroughly review these scripts so it's also possible they won't actually break (maybe only gadgets would?). At any rate, I've no issue with making Morebits ES6-only, but I don't think we should cause breakage, even for personal JS, without some sort of heads up.

In short, if you all are confident, I am too :)

**Comment by MusikAnimal** [comment]
> The issue would be that we would need a separate lint config for morebits, and I don't think eslint supports file-specific override for ecmaVersion (though I suppose we could put morebits in a subdir and put another .eslintrc in there).

Frankly, I've always wondered why Morebits wasn't set up that way. It seems odd to have the source files for a project of this size in the parent directory. A `lib/` directory or something might be more appropriate, given the intention of Morebits. I'm not involved or particularly interested in maintaining Morebits as a standalone library, but if others wish it to be that way, it might even make sense as a separate repository. Overall I think the long-term future of Twinkle should look to OOUI and Codex, but I digress.

**Comment by NovemLinguae** [comment]
Per [this onwiki discussion](https://en.wikipedia.org/wiki/Wikipedia_talk:Twinkle#Deploy), Morebits was set to |requiresES6 today.
